src/pybind/mgr/cephadm/agent.py

```diff
@@ -51,6 +53,30 @@ def __init__(self, mgr: "CephadmOrchestrator") -> None:
         self.server_addr = self.mgr.get_mgr_ip()
         super(CherryPyThread, self).__init__(target=self.run)
 
+    def configure_cherrypy(self) -> None:
+        cherrypy.config.update({
+            'environment': 'production',
+            'server.socket_host': self.server_addr,
+            'server.socket_port': self.server_port,
+            'engine.autoreload.on': False,
+            'server.ssl_module': 'builtin',
+            'server.ssl_certificate': self.cert_tmp.name,
+            'server.ssl_private_key': self.key_tmp.name,
+        })
+
+        # configure routes
+        root = Root(self.mgr)
+        host_data = HostData(self.mgr)
+        d = cherrypy.dispatch.RoutesDispatcher()
+        d.connect(name='index', route='/prometheus/', controller=root.index)
+        d.connect(name='sd-config', route='/prometheus/sd-config', controller=root.get_sd_config)
+        d.connect(name='rules', route='/prometheus/rules', controller=root.get_prometheus_rules)
+        d.connect(name='host-data', route='/data', controller=host_data.POST,
+                  conditions=dict(method=['POST']))
+
+        conf = {'/': {'request.dispatch': d}}
+        cherrypy.tree.mount(None, "/", config=conf)
+
```

Comment:
I'm starting to think that rather than hard-coding the services here, we should perhaps have a service registry/locator somewhere. The truth is that Ceph already provides a service registry:
```py
import rados
cluster = rados.Rados(conffile='/ceph/build.ceph/ceph.conf')
cluster.connect()
cluster.service_daemon_register('prometheus', '3', metadata=dict(port='1234'))  # immutable metadata
cluster.service_daemon_update(dict(status='down'))  # mutable status
```
Registered services can be checked both from CLI (`ceph service dump`):
```json
{
    "epoch": 8,
    "modified": "2022-03-16T20:32:26.515789+0000",
    "services": {
        "prometheus": {
            "daemons": {
                "summary": "",
                "3": {
                    "start_epoch": 8,
                    "start_stamp": "2022-03-16T20:32:26.275015+0000",
                    "gid": 4392,
                    "addr": "172.20.0.2:0/3598778827",
                    "metadata": {
                        "arch": "x86_64",
...
                        "port": "1234"
                    },
                    "task_status": {}
                }
            }
        },
        "rgw": {
            "daemons": {
                "summary": "",
                "4202": {
                    "start_epoch": 2,
                    "start_stamp": "2022-03-16T20:08:45.445458+0000",
                    "gid": 4202,
                    "addr": "172.20.0.2:0/4156482303",
                    "metadata": {
                        "arch": "x86_64",
                        "ceph_release": "quincy",
                        "ceph_version": "ceph version 17.0.0-10923-gc1a4d9ce (c1a4d9ce496f0fad7e1b7428d40a8137ff58239a) quincy (dev)",
                        "ceph_version_short": "17.0.0-10923-gc1a4d9ce",
...
                        "realm_id": "",
                        "realm_name": "",
                        "zone_id": "ee42dd26-78af-4004-8efe-c0a6810738f5",
                        "zone_name": "default",
                        "zonegroup_id": "175444c7-7968-4d67-8425-7e46ffbdceda",
                        "zonegroup_name": "default"
                    },
                    "task_status": {}
                }
            }
        }
    }
}

```
And `ceph service status`:
```json
{
    "prometheus": {
        "3": {
            "status_stamp": "2022-03-16T20:34:26.281015+0000",
            "last_beacon": "2022-03-16T20:36:31.286757+0000",
            "status": {
                "status": "down"
            }
        }
    },
    "rgw": {
        "4202": {
            "status_stamp": "2022-03-16T20:36:32.025856+0000",
            "last_beacon": "2022-03-16T20:36:32.025856+0000",
            "status": {
                "current_sync": "[]"
            }
        }
    }
}
```
However this seems to be very Ceph specific (you can only register one service per Rados connection), but it could serve as inspiration... What do you think?

Reply:
Thank you very much for your suggestion. Having a service discovery in Ceph in general I think it's a good idea not only for Prometheus but for any other services. However I'd like to address this as a more generic feature and and keep it as separated effort/PR.